### ISSUE TITLE
Custom client

### ISSUE THREAD

**Issue by theofidry** [csarrazi/CsaGuzzleBundle#43]
I'm declaring my client as a service since I wish to inject some dependencies in it. My client declaration is:

```yaml
services:
    api.client:
        class: FrontBundle\Client\ApiClient
        arguments:
            - @router
            - http://localhost/api
        tags: [ { name: csa_guzzle.client } ]
```

However this throw the following error:

    PHP Fatal error:  Cannot use object of type Symfony\Component\DependencyInjection\Reference as array in /path/to/myProject/vendor/csa/guzzle-bundle/src/DependencyInjection/CompilerPass/MiddlewarePass.php on line 106

Any idea?

Edit:
* Using Symfony2.7 and the version `dev-master` of the bundle

**Comment by csarrazi** [comment]
Your problem comes from the fact that you are extending the client in a wrong way. FYI, the only way to inject middleware in the Guzzle client is to inject the middleware in a handler, which is then passed in the client's options array. See http://guzzle.readthedocs.org/en/latest/handlers-and-middleware.html

So instead of using constructor injection, you should rather use setter injection, by creating a `setRouter()` method, and using `call` to add a method call after instanciation. For the URI, you should use Guzzle's recommended method, which is to pass it in the configuration array.

E.g:

```yml
services:
    api.client:
        class: FrontBundle\Client\ApiClient
        arguments:
            - { uri: "http://localhost/api" }
        calls:
            - { method: setRouter, [ "@router" ]}
        tags: [ { name: csa_guzzle.client } ]
```

By the way, unless you explicitely want to use Guzzle 6.0, you may want to stick with a stable version (`^1.3`) of the bundle, instead of `dev-master`, which is not production-ready yet (there's even a warning at the top of the bundle's documentation). :)

This issue could be fixed by using reflexion, and introspect the class' constructor to detect whether there is an options argument in the client, but there will be other edge cases, so I don't think it's worth all the hastle. I'll add documentation to explain how to correctly extend the Guzzle client. :)

**Comment by csarrazi** [comment]
@theofidry Did you try anything?

**Comment by theofidry** [comment]
I'm currently trying, I get the following error:

    Warning: strpos() expects parameter 1 to be string, array given
500 Internal Server Error - ContextErrorException

```
in vendor/guzzle/guzzle/src/Guzzle/Parser/UriTemplate/UriTemplate.php at line 41
39.    public function expand($template, array $variables)
40.    {
41.        if ($this->regex == self::DEFAULT_PATTERN && false === strpos($template, '{')) {
42.            return $template;
43.        }
```

Btw just a minor erorr in the service declaration:

```yaml
services:
    api.client:
        class: FrontBundle\Client\ApiClient
        arguments:
            - { uri: "http://localhost/api" }
        calls:
            - [ setRouter, [ "@router" ] ]
        tags: [ { name: csa_guzzle.client } ]
```

or 

```yaml
services:
    api.client:
        class: FrontBundle\Client\ApiClient
        arguments:
            - { uri: "http://localhost/api" }
        calls:
            - { method: setRouter, arguments: [ "@router" ] }
        tags: [ { name: csa_guzzle.client } ]
```
But that's not the issue

**Comment by csarrazi** [comment]
My bad. Change the config to this:

```yml
services:
    api.client:
        class: FrontBundle\Client\ApiClient
        arguments:
            - { uri: "http://localhost/api" }
        calls:
            - [ setRouter, [ "@router" ]]
        tags: [ { name: csa_guzzle.client } ]
```

It should work better.

**Comment by csarrazi** [comment]
Ok. The second problem is the content of the Uri, it seems. Try to check what you have in the generated service, in `appDevDebugProjectContainer.php`.

**Comment by theofidry** [comment]
I did, it's not always easy to get the service declaration right so I guess it was a mistaked^^, but that gives the error above with the proper declaration

**Comment by theofidry** [comment]
I think it's `base_uri` and not `uri`, at least according to the [PHPDoc in Guzzle client](https://github.com/guzzle/guzzle/blob/master/src/Client.php#L38), but still gives the same error.

The generated service:
```php
protected function getApi_ClientService()
    {
        $a = \GuzzleHttp\HandlerStack::create();
        $a->push($this->get('csa_guzzle.middleware.stopwatch'), 'stopwatch');
        $a->push($this->get('csa_guzzle.middleware.history'), 'history');
        $a->push($this->get('csa_guzzle.middleware.logger'), 'logger');

        $this->services['api.client'] = $instance = new \FrontBundle\Client\ApiClient(array('base_uri' => 'http://localhost/', 'handler' => $a));

        $instance->setRouter($this->get('router'));

        return $instance;
    }
```

**Comment by theofidry** [comment]
If you wish to have an insight of my class basically it does not do much it just allows to pass route names instead of just URI, so if I do a get for the route `list_users` it will call the parent class (the Guzzle Client) get with `/api/users`.

**Comment by csarrazi** [comment]
Hum, I see nothing wrong with the service itself.
It seems the new error is linked to the call you are making, rather than the service.

How are you calling your client to make your call?

```php
$this->get('api.client')->get('hello_world');
```

**Comment by theofidry** [comment]
Exactly. And I've check the `Guzzle::get()` call the parameters passed are fine.

**Comment by csarrazi** [comment]
Could you show your client's get method?

**Comment by theofidry** [comment]
```php
<?php
namespace FrontBundle\Client;

use Guzzle\Common\Collection;
use Guzzle\Common\Exception\RuntimeException;
use Guzzle\Http\Client as GuzzleClient;
use Symfony\Bundle\FrameworkBundle\Routing\Router;

class ApiClient extends GuzzleClient implements ApiClientInterface
{
    /**
     * @var Router
     */
    private $router;

    /**
     * @param Router $router  Component used to generate URI from route names
     */
    public function setRouter(Router $router)
    {
        $this->router = $router;
    }

    public function get($uriOrRouterName = null, $token = null, $options = [])
    {
        return parent::get($this->getUri($uriOrRouterName), $this->extractHeaders($token, $options), $options);
    }

    // other methods

    private function extractHeaders($token = null, array &$options = [])
    {
        // Extract header from options
        $headers = [];
        if (isset($options['headers'])) {
            $headers = $options['headers'];
            unset($options['headers']);
        }

        // Add authorization token
        if (null !== $token) {
            $headers['authorization'] = sprintf('Bearer %s', $token);
        }

        return $headers;
    }

    private function getUri($uriOrRouterName)
    {
        return (false !== strpos($uriOrRouterName, '/')) ? $uriOrRouterName : $this->router->generate($uriOrRouterName);
    }
}
```

**Comment by theofidry** [comment]
Can you reproduce the bug? Otherwise I can try to give you more data regarding the stacktrace.

**Comment by theofidry** [comment]
Otherwise is on GitHub anyway: [theofidry/Incipio](https://github.com/theofidry/Incipio).

**Comment by csarrazi** [comment]
Je vais regarder demain matin. Là, l'heure se fait un peu tardive ! :)

**Comment by theofidry** [comment]
Duh, so stupid. The I was extended the wrong Guzzle client. Still, as I needed to change my signatures (I have my own ClientInterface which does not extend the Guzzle one) it was not possible to extend Guzzle client. So I declared a client as recommended in your config and my true client is in fact an adaptor based on the client declared via your bundle. Request are registered fine so, so far so good.
